cmd/gpu-kubelet-plugin/device_health.go

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+	"k8s.io/klog/v2"
+)
+
+const (
+	FullGPUInstanceID uint32 = 0xFFFFFFFF
+)
+
+type deviceHealthMonitor struct {
+	nvmllib         nvml.Interface
+	eventSet        nvml.EventSet
+	unhealthy       chan *AllocatableDevice
+	cancelContext   context.CancelFunc
+	uuidToDeviceMap map[string]*AllocatableDevice
+	wg              sync.WaitGroup
+}
+
+func newDeviceHealthMonitor(ctx context.Context, config *Config, allocatable AllocatableDevices, nvdevlib *deviceLib) (*deviceHealthMonitor, error) {
```

Comment:
This doesn't just create the health monitor, it also starts it. Why not separate construction from actually starting the monitor? This will allow us to more easily swap out monitors (e.g. with NVSentinel) once we have those ready.

Reply:
QUestion: What functionality is required for a `DeviceHealthMonitor` from the perspective of the driver. Would returning the following `interface` here make sense:

```go
type DeviceHealthMonitor interface {
    Start() error
    Stop()
    Unhealthy() <-chan *AllocatableDevice
}
```

Reply:
Great discussion! Let's leave this thread open. I think we can get back to this after merging the patch. Vetos?